### ISSUE TITLE
Set a timeout to a request

### ISSUE THREAD

**Issue by macistador** [algolia/algoliasearch-client-swift#135]
Hi 

I can't find a way to set a timeout in ms which would cancel a request, is there an alternative ?
I would like to do something else if a request does not respond until a period and avoid thus to handle a response I don't need in this case anymore (it would simplify the code).

Thanks

**Comment by clement-leprovost** [comment]
**TL;DR:** There is currently no way to set a "hard" timeout for requests. If you need, you can always set a timer and cancel the request when it fires.

**Long story:** There are global timeouts at the `Client` level:

1. `searchTimeout` for search requests
2. `timeout` for all other types of requests

... but these are for individual network requests; then the **retry logic** is applied on top of them: if a request hits a timeout, it will be retried on another host, until the list of hosts is exhausted. Also, the timeout will be automatically increased to accommodate for potentially slow network conditions. All in all, the elapsed time can be rather long. If timeouts are hit on all hosts, then the request results in an error.

Maybe we should consider adding a feature to set a "hard timeout" on every request. But again, hitting the timeout would result in an error, not cancellation. Also, this would have to be specified at the `Client` level (i.e. for all requests), otherwise it would likely result in a breaking change (or would require to duplicate *every* request method, which is not a viable option).

@macistador What do you think?

**Comment by FJuteau** [comment]
Hi

@clement-leprovost I, currently, am working on a time out too and I would love to try to set a timer and cancel the requests but I can't find a way to cancel all requests from an index like I use to in previous version : index.cancelPreviousRequests() 
Client version : 3.7

Thanks

**Comment by macistador** [comment]
@clement-leprovost  Ok I understand your point on implementing the timeout at the client level, it makes sense and would be just fine in our case

**Comment by clement-leprovost** [comment]
@FJuteau There is indeed no longer any way to cancel all previous requests at once, but the `Client` and `Index` classes return `Request` (basically `Operation`) instances that can be used for cancelling. We can consider re-adding a "cancel all" method, but we would need to understand the use case better to make sure it's not redundant with returning individual requests. Can you tell us more about why you need to cancel them in batch?

@macistador A hard time-out is definitely something that we have started investigating. It's just still unclear whether it should be added at the client level, or in a higher-level helper library. I can't provide you with any ETA yet, so if you need it quickly, then I advise you to build your own helper by keeping a reference to the returned `Request` object and cancelling it after a timer.

**Comment by clement-leprovost** [comment]
Awakening this issue that has been sleeping for quite some time...

The [InstantSearch Core](https://github.com/algolia/instantsearch-core-swift) library, released a few months ago, allows [cancelling requests](https://community.algolia.com/instantsearch-core-swift/reference/Classes/Searcher.html#/s:FC17InstantSearchCore8Searcher21cancelPendingRequestsFT_T_) on a given `Searcher` (i.e. basically an index). The searcher also takes care of discarding responses to obsolete requests (i.e. when a more recent response has arrived).

Note that there is still no way to set a hard time out on every request, but I don't know if it's really needed given those higher-level facilities. @macistador @FJuteau What do you think?
